apps/emails-and-messages/src/modules/sendgrid/configuration/sendgrid-config-input-schema.ts

```diff
@@ -89,7 +89,19 @@ export type SendgridUpdateEvent = z.infer<typeof sendgridUpdateEventSchema>;
 
 export const sendgridUpdateEventArraySchema = z.object({
   configurationId: z.string(),
-  events: z.array(sendgridConfigurationEventSchema),
+  events: z
+    .array(sendgridConfigurationEventSchema)
+    /*
+     * Pass the validation if all the events are in one of two states:
+     * 1. Inactive
+     * 2. Active and have a template
+     */
+    .refine(
+      (data) => data.every((event) => event.active === false || (event.active && event.template)),
+      {
+        message: "All active events must have assigned template.",
+      }
+    ),
```

Comment:
I was big on `zod` doing more than just parsing the schema, but after some nasty bugs, I changed my mind.

I think once we cross that line it will be difficult to establish what "schemas" even do in our apps. By looking at the variable name `sendgridUpdateEventArraySchema`, you can't possibly guess there is logic there. Not sure about the structure of this app, but in Taxes I have a service for managing the provider configuration, and I would put that logic there.

Reply:
Initially, I hoped to put this validation on `sendgridConfigurationEventSchema`, but it turns out "refined" schemas cant be used to compose (merge, extent, etc) :/ 
Most likely "I'm using zod the wrong way" with my approach.

Anyway, putting validation here has the advantage of client-side validation - `zodResolver` Only submits the form if this requirement is met.

Reply:
I dont see any problem here. `refine` is an API for custom validation fn. Zod is validation library. IMO its the right way to put it. 

However, it should be tested

Reply:
Tests have been added.